intermediate_source/transformer_building_blocks.py

```diff
@@ -0,0 +1,641 @@
+                                                                padded_query,
+                                                                padded_query,
+                                                                key_padding_mask=src_key_padding_mask,
+                                                                need_weights=False,
+                                                                attn_mask=attn_mask,
+                                                                is_causal=True)
+
+print(f"{padded_time=:.5f}, padded_peak_memory={padded_peak_memory/1e9:.2f} GB")
+print(f"{nested_time=:.5f}, nested_peak_memory={nested_peak_memory/1e9:.2f} GB")
+print("Difference between vanilla and nested result", (padded_result - padded_nested_result).abs().max().item())
+print(f"Nested speedup: {(padded_time/nested_time):.2f}")
+print(f"Nested peak memory reduction {((padded_peak_memory - nested_peak_memory)/1e9):.2f} GB")
+
+######################################################################################
+# We can also see the same for backward pass
+# padding-specific step: remove output projection bias from padded entries for fair comparison
+for i, entry_length in enumerate(sentence_lengths):
+    padded_result[i, entry_length:, :] = 0.0
+
+_, padded_bw_time, padded_bw_peak_mem = benchmark(lambda : padded_result.sum().backward())
+_, nested_bw_time, nested_bw_peak_mem = benchmark(lambda : padded_nested_result.sum().backward())
+
+print(f"{padded_bw_time=:.5f}, padded_bw_peak_mem={padded_bw_peak_mem/1e9:.2f} GB")
+print(f"{nested_bw_time=:.5f}, nested_bw_peak_mem={nested_bw_peak_mem/1e9:.2f} GB")
+print(f"Nested backward speedup: {(padded_bw_time/nested_bw_time):.2f}")
+print(f"Nested backward peak memory reduction {((padded_bw_peak_mem - nested_bw_peak_mem)/1e9):.2f} GB")
+
+print("Difference in out_proj.weight.grad", (mha_layer.out_proj.weight.grad - vanilla_mha_layer.out_proj.weight.grad).abs().max().item())
+print("Difference in packed_proj.weight.grad", (mha_layer.packed_proj.weight.grad - vanilla_mha_layer.in_proj_weight.grad).abs().max().item())
+print("Difference in out_proj.bias.grad", (mha_layer.out_proj.bias.grad - vanilla_mha_layer.out_proj.bias.grad).abs().max().item())
+print("Difference in packed_proj.bias.grad", (mha_layer.packed_proj.bias.grad - vanilla_mha_layer.in_proj_bias.grad).abs().max().item())
+
+##################################################################################
+# GPT-style layer
+# ---------------
+# A basic GPT-style transformer layer consistst of a causal self-attention layer
+# followed by a feed-forward network (FFN) with skip connections. Implementing
+# this is fairly straightforward using the ``MultiheadAttention`` layer above and
+# is actually the same as an ``nn.TransformerEncoderLayer`` with ``is_causal=True``.
+
+# We  demonstrate examples of implementing the rest of the nn layers
+# `here <https://github.com/mikaylagawarecki/temp>`_ but omit that from this
```

Comment:
I assume the name / hosting for this repo will change before publishing the tutorial?

Reply:
name yes, hosting, perhaps we can make that a followup -- I don't think we want to host this at pytorch/<foo> yet as that lends a lot of "official"-ness to this repo which I don't want it to have just yet